metagraph/src/graph/representation/base/dbg_wrapper.hpp

```diff
@@ -0,0 +1,325 @@
+        return graph_->has_single_incoming(node);
+    }
+
+    virtual node_index kmer_to_node(std::string_view kmer) const override {
+        return graph_->kmer_to_node(kmer);
+    }
+
+    virtual void call_outgoing_kmers(node_index kmer,
+                                     const OutgoingEdgeCallback &callback) const override {
+        graph_->call_outgoing_kmers(kmer, callback);
+    }
+
+    virtual void call_incoming_kmers(node_index kmer,
+                                     const IncomingEdgeCallback &callback) const override {
+        graph_->call_incoming_kmers(kmer, callback);
+    }
+
+    // Check whether graph contains fraction of nodes from the sequence
+    virtual bool find(std::string_view sequence, double discovery_fraction = 1) const override {
+        return graph_->find(sequence, discovery_fraction);
+    }
+
+    virtual bool operator==(const DeBruijnGraph &other) const override = 0;
+
+    virtual const std::string& alphabet() const override { return graph_->alphabet(); }
+    virtual void print(std::ostream &out) const override { graph_->print(out); }
+
+    // Call all nodes that have no incoming edges
+    virtual void call_source_nodes(const std::function<void(node_index)> &callback) const override {
+        graph_->call_source_nodes(callback);
+    }
+
+    virtual const DeBruijnGraph& get_base_graph() const override final {
+        return graph_->get_base_graph();
+    }
+
+  protected:
+    std::shared_ptr<Graph> graph_ptr_;
+    std::shared_ptr<const Graph> graph_;
+};
+
+template <typename Graph = DeBruijnGraph>
+class DBGNodeModifyingWrapper : public DBGWrapper<Graph> {
+  public:
+    typedef typename Graph::node_index node_index;
+    typedef typename Graph::CallPath CallPath;
+    typedef typename Graph::OutgoingEdgeCallback OutgoingEdgeCallback;
+    typedef typename Graph::IncomingEdgeCallback IncomingEdgeCallback;
+
+    template <typename... Args>
+    explicit DBGNodeModifyingWrapper(Args&&... args)
+          : DBGWrapper<Graph>(std::forward<Args>(args)...) {}
+
+    virtual ~DBGNodeModifyingWrapper() {}
+
+    virtual void map_to_nodes(std::string_view sequence,
+                              const std::function<void(node_index)> &callback,
+                              const std::function<bool()> &terminate
+                                  = [](){ return false; }) const override = 0;
+
+    virtual void map_to_nodes_sequentially(std::string_view sequence,
+                                           const std::function<void(node_index)> &callback,
+                                           const std::function<bool()> &terminate
+                                               = [](){ return false; }) const override = 0;
+
+    virtual void
+    adjacent_outgoing_nodes(node_index node,
+                            const std::function<void(node_index)> &callback) const override = 0;
+
+    virtual void
+    adjacent_incoming_nodes(node_index node,
```

Comment:
what is this class about?

Reply:
I'll add a comment. The idea here is that if the graph modifies the external view of the graph (adds/removes nodes), then these newly-made pure virtual methods are the bare minimum that need to be implemented. For the rest of the methods, this uses the `DeBruijnGraph` versions instead of the template type's by default.

Reply:
But how many classes will inherit from this? Isn't this needed only for canonical_dbg?

Reply:
`MaskedDeBruijnGraph` and `RCDBG` (#333) inherit from this

Reply:
while `DBGSuccinctCache` inherits from `DBGWrapper`

Reply:
why do you need to add/remove virtual nodes in RCDBG?

Reply:
what is the point of this class, anyway? that you defined methods like
```
    virtual bool has_single_outgoing(node_index node) const override {
        return DeBruijnGraph::has_single_outgoing(node);
    }
```
?

But isn't that already defined the same way in DBGWrapper?

Reply:
The wrapper reverses the direction of all traversal and returns the reverse complement for its node labels, so its external view is different from the underlying graph (using the same node indices).

Reply:
> what is the point of this class, anyway? that you defined methods like
> 
> ```
>     virtual bool has_single_outgoing(node_index node) const override {
>         return DeBruijnGraph::has_single_outgoing(node);
>     }
> ```
> 
> ?
> 
> But isn't that already defined the same way in DBGWrapper?

`DBGWrapper` calls all of its methods from the underlying graph, whereas `DBGNodeModifyingWrapper` tends to call the `DeBruijnGraph` versions

Reply:
This definitely sounds too complex, even if it makes sense.

Is it possible to do things without this class?

Reply:
The approaches I see are:
1. have each method run `graph_->myop(...)` and trust the child class to override everything relevant
2. have each method run `DeBruijnGraph::myop(...)` and trust the child class to override based on efficiency concerns (e.g., manually checking if the type is DBGSuccinct)
3. make every method which has non-negligible run time pure virtual
4. some combination of the above

Reply:
The child must override everything relevant in any case. I don't see how making these intermediate classes may help.

Reply:
Not exactly. In `CanonicalDBG`, for example, methods like `has_single_outgoing`, 'has_multiple_outgoing`, `traverse(node_index, const char*, ...)`, etc. can use the default implementations from `DeBruijnGraph`. With this structure, one doesn't have to manually override them.

Reply:
I cleaned up the two wrappers now and added descriptions, so it should be easier to understand now.

Reply:
Just always re-set those to `DeBruijnGraph::...`

Reply:
But if everything needs to be reset anyway, there's no point in having a parent wrapper class.

Reply:
the point of the class is to avoid writing forwards like `return graph_.traverse(...);` in every wrapper.

This abstract parent delegator-wrapper should be a single thing. We don't need 2.

Reply:
Yes, but then sometimes you want to delegate to the base graph's method, and other times to `DeBruijnGraph`'s. If only one of them is implemented, then you need to spend time writing delegations for the other one.

Reply:
My point is, let's make this wrapped graph with overloaded methods behave exactly the same way as if this was overloaded for the base DeBruijnGraph. In this case, things will be natural and you don't need to check every second the code to double-check how things work. Because this is the ultimate goal of the interfaces -- they should assume the most natural and simple behavior you could expect.

Reply:
And even having the name `DBGNodeModifyingWrapper` clearly doesn't satisfy this condition.

Reply:
So I'm guessing you want it to look something like this? 0085837a56e0faaf22f40ab89408c1960c1bf7a6

The issue I have with this implementation is that there are so many methods in `DeBruijnGraph` that it's really easy to forget what needs to be overloaded and make mistakes.

With `DBGSuccinctCached`, for example, one would have to look through `sequence_graph.hpp` and find which methods have default implementations for `DeBruijnGraph` so they can be overloaded.

Reply:
Think of one of those wrappers: CanonicalDBG, etc.
The purpose of this abstract wrapper is to avoid forwarding the calls duplicated among all of them.

So just define those forwards in here, and keep the rest pure abstract, so they are defined in the children.

Reply:
But it might also be that this idea with writing common interface class for wrappers is not that good after all and just rewriting all methods in all wrappers is better and more straighforward.

Reply:
I think it's still useful to have an abstract wrapper parent to deal with stuff like `get_k`, `alphabet`, etc, and also constructors.

I've reduced the class down to a minimum so that anyone who uses it has to be mindful of what should be implemented in derived classes, while not having to deal with the trivial stuff that's implemented in the parent.